File: src/main/java/fr/uiytt/ageofempire/gui/villagergui/TeamGui.java
package fr.uiytt.ageofempire.gui.villagergui;

import fr.minuskube.inv.ClickableItem;
import fr.minuskube.inv.SmartInventory;
import fr.minuskube.inv.content.InventoryContents;
import fr.minuskube.inv.content.InventoryProvider;
import fr.uiytt.ageofempire.AgeOfEmpire;
import fr.uiytt.ageofempire.game.GameManager;
import fr.uiytt.ageofempire.game.GameTeam;
import fr.uiytt.ageofempire.utils.PlayerFromUUIDNotFoundException;
import fr.uiytt.ageofempire.utils.Utils;
import net.md_5.bungee.api.ChatColor;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public class TeamGui implements InventoryProvider {

  private SmartInventory inventory;

  public TeamGui() {
    inventory = SmartInventory.builder()
            .id("AOE_Team")
            .size(3, 9)
            .title("Team")
            .provider(this)
            .manager(AgeOfEmpire.getInvManager())
            .build();
  }

  @Override
  public void init(Player player, InventoryContents contents) {
    contents.fillBorders(ClickableItem.empty(Utils.newItemStack(Material.GRAY_STAINED_GLASS_PANE, ChatColor.GRAY + "", List.of("") )));
  }

  @Override
  public void update(Player player, InventoryContents contents) {
    int[] slots = new int[]{3, 5, 1, 7};
    int i = 0;
    for(GameTeam team : GameManager.getGameInstance().getGameData().getTeams()) {
      contents.set(1, slots[i], ClickableItem.of(
          Utils.newItemStack(
              team.getColor().getBanner(),
              team.getColor().getChatColor() + team.getName(),
              loreBuilder(team),
              1),
          event -> {
            if(addPlayer(team, player)) {
              player.sendMessage("vous avez été ajouté à la team");
            }
          }
      ));
      i++;
    }

  }

  /**
   * For a given team, return a list of player and available slots in the team to be shown in the lore
   * @param team A team which may contains player
   * @return a list of players and slots for this team as Lore {@see List<String>}
   */
  private List<String> loreBuilder(GameTeam team) {
    List<String> lore = new ArrayList<>();
    for(int i=0;i<8;i++) {
      if(team.getPlayersUUIDs().size() - 1 >= i) {
        UUID playerUUID = team.getPlayersUUIDs().get(i);
        Player player = Bukkit.getPlayer(playerUUID);
        if(player == null) {
          team.removePlayer(playerUUID);
          return loreBuilder(team);
        }
        lore.add(team.getColor().getChatColor() + "- " + player.getName());
      } else {
        lore.add(team.getColor().getChatColor() + "- _______");
      }
    }
    return lore;
  }

  /**
   * Try to add a player to a team
   * @param team the team to join
   * @param player the player that want to join
   * @return true if the player was added
   */
  private boolean addPlayer(GameTeam team, Player player) {
    if(team.getPlayersUUIDs().size() >= 8) {
      return false;
    }
    GameTeam previousTeam = GameManager.getGameInstance().getGameData().getPlayersTeam().get(player.getUniqueId());
    try {
      if(previousTeam != null) {
        previousTeam.removePlayer(player.getUniqueId());
      }
      team.addPlayer(player.getUniqueId());
    } catch (PlayerFromUUIDNotFoundException e) {
      e.printStackTrace();
      return false;
    }
    return true;
  }

  public void openGUI(Player player) {
    inventory.open(player);
  }
}

File: src/main/kotlin/fr/uiytt/ageofempire/gui/TeamGui.kt
package fr.uiytt.ageofempire.gui

import fr.minuskube.inv.ClickableItem
import fr.minuskube.inv.SmartInventory
import fr.minuskube.inv.content.InventoryContents
import fr.minuskube.inv.content.InventoryProvider
import fr.uiytt.ageofempire.AgeOfEmpire
import fr.uiytt.ageofempire.game.GameTeam
import fr.uiytt.ageofempire.game.getGameManager
import fr.uiytt.ageofempire.game.getPlayerTeam
import fr.uiytt.ageofempire.utils.PlayerFromUUIDNotFoundException
import fr.uiytt.ageofempire.utils.Utils.newItemStack
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player

class TeamGui : InventoryProvider {
    private val inventory: SmartInventory = SmartInventory.builder()
        .id("AOE_Team")
        .size(3, 9)
        .title("Team")
        .provider(this)
        .manager(AgeOfEmpire.invManager)
        .build()

    override fun init(player: Player, contents: InventoryContents) {
        contents.fillBorders(
            ClickableItem.empty(
                newItemStack(Material.GRAY_STAINED_GLASS_PANE, ChatColor.GRAY.toString() + "", listOf("")))
        )
    }

    override fun update(player: Player, contents: InventoryContents) {
        val slots = intArrayOf(3, 5, 1, 7)
        for ((i, team) in getGameManager().gameData.teams.withIndex()) {
            contents[1, slots[i]] = ClickableItem.of(
                newItemStack(team.color.banner, team.color.chatColor.toString() + team.name, loreBuilder(team), 1)
            ) { if (addPlayer(team, player)) { player.sendMessage("vous avez été ajouté à la team") } }
        }
    }

    /**
     * For a given team, return a list of player and available slots in the team to be shown in the lore
     *
     * @param team A team which may contains player
     * @return a list of players and slots for this team as Lore {@see List<String>}
     */
    private fun loreBuilder(team: GameTeam): List<String?> {
        val lore: MutableList<String?> = ArrayList()
        for (i in 0..7) {
            if (team.playersUUIDs.size - 1 >= i) {
                val playerUUID = team.playersUUIDs[i]
                val player = Bukkit.getPlayer(playerUUID)
                if (player == null) {
                    team.removePlayer(playerUUID)
                    return loreBuilder(team)
                }
                lore.add(team.color.chatColor.toString() + "- " + player.name)
            } else {
                lore.add(team.color.chatColor.toString() + "- _______")
            }
        }
        return lore
    }

    /**
     * Try to add a player to a team
     *
     * @param team   the team to join
     * @param player the player that want to join
     * @return true if the player was added
     */
    private fun addPlayer(team: GameTeam, player: Player): Boolean {
        if (team.playersUUIDs.size >= 8) {
            return false
        }
        val previousTeam: GameTeam? = player.uniqueId.getPlayerTeam()
        try {
            previousTeam?.removePlayer(player.uniqueId)
            team.addPlayer(player.uniqueId)
        } catch (e: PlayerFromUUIDNotFoundException) {
            e.printStackTrace()
            return false
        }
        return true
    }

    fun openGUI(player: Player?) {
        inventory.open(player)
    }
}